### DISCUSSION TITLE
Add support for absolute trackpad input

### DISCUSSION THREAD

**Discussion by KevZhi** [ppy/osu#12988] [Ideas]
I'm a MacBook user and playing osu! with MacBook Built-in trackpad. 
![imb_qdgu2f](https://user-images.githubusercontent.com/32162677/42632108-bc1bc188-860e-11e8-8cd2-5440079fe697.GIF)

There's an app Inklet can map real trackpad touch location to gameplay screen just like tablet. But this app have some** input delay**, also needs to manually quit and start mapping mode when selecting songs/mods or playing.

There's some document from apple which may help

https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/EventOverview/TrackingAreaObjects/TrackingAreaObjects.html#//apple_ref/doc/uid/10000060i-CH8-SW3

and new synaptics trackpad on widows laptop support getting raw trackpad touch location.

I hope this feature will be a built-in feature in lazer, thanks anyway.

some "light" players using laptop who doesn't want to have a tablet like pro players, just simply enjoy the game, will like this feature, also, playing with trackpad is way more portable which in the occasion just like you are in airport waiting for a flight, enjoy with a game.

**Comment by ticotaco72** [comment]
It would be possibly be hard to do with older synaptics hardware, non synaptics hardware and on Linux (everything should be cross platform). But maybe possible, i don't know this topic.

**Comment by KevZhi** [comment]
@ticotaco72 syncs.exe is used for get absolute position on trackpad on windows. 
My opinion is that trackpad absolute mapping should be a external hardware support like wii taiko drum support in osu! so not needed to be cross platform. 
Actually Synaptics has a large market share of PC trackpad. Also, as long as the laptop can run osu in acceptable quality, I don't think it will have too much possibilities comes with an old shit trackpad.

anyhow, I think that feature will enhance the experience of the players don't want to buy tablets and just want to do some relax game on their laptop.

I think this features can be planned after  goes stable, not urgent.

**Comment by ticotaco72** [comment]
@KevZhi Not everything old hardware is shit. 
And how long is available getting raw position in synaptics API?
What's definition of old? 2 years, 10 years
And why we want to say "people with older hardware or exotic hardware or Linux are worse". Lazer is build for cross-platform

**Comment by KevZhi** [comment]
@ticotaco72 
it's already some softwares here and API is open. its defiantly not a huge work.
since windows 7 going out, laptops starts to support multi touch gestures on trackpad. a large amount of  main stream laptops supports now.

That feature I mentioned is NOT a necessary to CrossPlatform, it's not a main function. You can either choose to buy a tablet or practice on mouse, or struggle with trackpad without using absolute positioning. 

That's not affecting on game playing. 
osu! just provided us ways to play with all kind of normal input device .

it's just an external support to a certain kind device, just like wii taiko drum does.
who has that kinda device, who can enjoy a convenient way playing osu! without buy additional hardware
and a large amount of player using laptop has that device.

so why not?

I'm absolutely not meaning that people with old hardware or linux are worse.

As I'm a Chinese user. old osu! Mac(wine package)never supported Chinese characters display, I don't feel any bad about it. After macOS 10.13 release, I can't even play osu because of  compatibility problems of wine for serval month or so until new version comes out.

Cuz I know Mac is for work, not for playing games, linux also, if anything happens, it's in expectation
.

If I'm a unsupported one I'd rather install a windows or buy a new pc or tablet to enjoy that , instead of asking osu! to add support to my device.

if I can do that kind of code I'm very glad to contribute, I think there will be guys intrested in

**Comment by peppy** [comment]
This is definitely something we want to support at some point in the future. It shouldn't be too hard to implement as an `InputHandler` if you'd like to give it a shot.

**Comment by KevZhi** [comment]
@peppy thanks a lot  for putting it on the list! 


https://gist.github.com/chamons/6efc2bc14e9e40b33a1d77826a0673bc ,will work on Mac maybe?
Actually sorry for that my coding skills can't supports me contribute to it.

**Comment by edgao** [comment]
@peppy is there an example of how to implement an InputHandler? The only relevant files I've found in the repo seem to be related to replays. I've also never done C# before so this will be a bit of an adventure :P

**Comment by edgao** [comment]
aha, found https://github.com/ppy/osu-framework/blob/master/osu.Framework/Input/Handlers/InputHandler.cs. I'll dig around and see if I can make sense of how this would look.

**Comment by peppy** [comment]
yep, best to start by making this work on the framework, completely separate from osu!. feel free to jump on discord and ask if you have any questions regarding the input hierarchy.

**Comment by AlexGuo1998** [comment]
Hi devs, recently I'm experimenting with "Precision Touchpad" on Windows.
It seems like recent multipoint touchpads that Windows PC supports are using HID protocol. With raw input and HID API, we can read absolute XY values.
I wrote a little demo (in C) and will later upload to GitHub, but I don't know much about C# and dotnet.
![demo screenshot](https://user-images.githubusercontent.com/25975864/72144633-bcfc7880-33d3-11ea-9c3b-92f319dfdf46.png)
Is there anyone working on this now? Maybe I can provide some help.

Docs from Microsoft:
* [Precision Touchpad Implementation Guide](https://docs.microsoft.com/en-us/windows-hardware/design/component-guidelines/precision-touchpad-implementation-guide)
* [Raw Input](https://docs.microsoft.com/en-us/windows/win32/inputdev/raw-input)
* [HID](https://docs.microsoft.com/en-us/windows-hardware/drivers/hid/)

**Comment by peppy** [comment]
There is ongoing work to add multitouch support, but no one is specifically adding precision trackpad support yet. You are welcome to implement support (at osu-framework, not here).

**Comment by AlexGuo1998** [comment]
OK. I'll take a look there.

**Comment by edgao** [comment]
sorry folks, turns out an OS-specific input method is a terrible way to learn a new programming language. I haven't made any real progress on the OSX absolute positioning feature - @AlexGuo1998 if you want to pick that up as well it's all yours; I might poke at it a little more but likely won't have time to do much with it

**Comment by mcendu** [comment]
Should we enable "Touch Device" mod when such type of input is detected to be in use?

**Reply by TheFantasticLoki** [reply]
IMO I don't think that is needed since it's not like your screen is on your touchpad

**Comment by smoogipoo** [comment]
Not a discussion that needs to be had right now.

**Comment by izzues** [comment]
Hey! Any updates on Precision Touchpad support?

**Comment by smoogipoo** [comment]
None yet.

**Comment by samanft** [comment]
Any updates?

**Reply by peppy** [reply]
See https://github.com/ppy/osu/discussions/12988#discussioncomment-791173.

**Reply by samanft** [reply]
damn 😭😭😭 thanks for the quick response though

**Comment by lokxii** [comment]
I wrote a command line utility for Mac that does absolute tracking on trackpad and works kinda well on Osu stable (wineskin). I haven't tested it on lazer. It is written in C that uses Apple's private framework Multitouch support to get raw trackpad events and hook callback to it. Then the utility listens to system wide mouse moved events and emits new mouse moved events with the mapped coordinate.

https://github.com/9tin9tin9/Mac-trackpad-mapper

It may be a temporary solution for Mac users.

Update: I have wrapped the utility into a status bar app that is more easy to use.

**Reply by tuxedoperson** [reply]
could you help me? i want to use this but i have no coding experience, and have no idea how to build or compile this.

**Reply by frenzibyte** [reply]
You want to ask the owner of that repository instead.

**Reply by tuxedoperson** [reply]
> You want to ask the owner of that repository instead.

they are the owner of the repository

**Reply by frenzibyte** [reply]
I mean, away from this discussion thread since it's pretty much off-topic.

**Comment by lin-raymond-dev** [comment]
Hi! I know it's been a few years, but I just want to ask if this feature is still on the roadmap for lazer? I think this feature would be super convenient to play Osu on the go with my laptop. I would love to know if this is still being considered? 

Thanks for taking the time :)

**Reply by AlexGuo1998** [reply]
I made a proof-of-concept implementation last year, at https://github.com/ppy/osu-framework/pull/6542
However it seems the focus wasn't there, so it was neither merged nor continue developed :(

**Comment by Yahddyyp** [comment]
I made something for this exact reason for macbook trackpads and it should work (i don't know how viable it is as i am way to bad at osu to test it). It uses apple's own frameworks to make it work 

https://github.com/Yahddyyp/osu-thing
